### PR TITLE
Write lobby filetimes to google sheet

### PR DESCRIPTION
**Pull request by jakobhellermann**
also includes the changes from #2 so reviewing that first should make this diff a bit easier to read

image

image

### CODE REVIEW

**Comment by Kataiser** [comment]
I'd rather sheet ID be given via a separate command, or prompted when registering a lobby project, rather than part of the already bloated register project command params

File: commands.py [hunk 1]
```diff
@@ -211,6 +213,21 @@ async def command_register_project(message: discord.Message, message_split: List
         await message.channel.send(f"Github app instllation cannot access the repo.")
         return
 
+    # verify lobby sheet
+    if lobby_sheet_cell.lower() != 'y' and is_lobby.lower() == 'n':
+        await utils.report_error(client, f"Cannot add lobby sheet '{lobby_sheet_cell}' to a non-lobby project.")
+        await message.channel.send("Cannot add lobby sheet to a non-lobby project.")
+        return
+    (spreadsheet_id, _, cell) = lobby_sheet_cell.partition('/')
```
Comment (Kataiser) [inline commands.py]: These parenthesis are redundant

File: commands.py [hunk 2]
```diff
@@ -211,6 +213,21 @@ async def command_register_project(message: discord.Message, message_split: List
...
+
+    try:
+        spreadsheet.check_write_permission(spreadsheet_id, cell)
+    except Exception as e:
```
Comment (Kataiser) [inline commands.py]: Tiniest nitpick ever but I'd prefer the exception to be named `error`

File: commands.py [hunk 3]
```diff
@@ -634,6 +653,7 @@ async def command_projects(message: discord.Message):
                f"\nImprovement channel: <#{project['project_id']}>" \
                f"\nAdmin{plural(admins)}: {', '.join(admins)}" \
                f"\nIs lobby: `{project['is_lobby']}`" \
+               f"\nLobby Sheet: `{project['lobby_sheet_cell']}`" \
```
Comment (Kataiser) [inline commands.py]: This should be a markdown link, and not shown if not a lobby project

File: spreadsheet.py [hunk 4]
```diff
@@ -383,6 +383,54 @@ def read_sheet(cell_range: str, multiple_rows=False):
         utils.log_error()
         raise SheetReadError
 
+def write_sheet(spreadsheet: str, cell_range: str, new_values: List[List[Union[int, str]]]):
```
Comment (Kataiser) [inline spreadsheet.py]: `list[list[Union[int, str]]]`

File: spreadsheet.py [hunk 5]
```diff
@@ -383,6 +383,54 @@ def read_sheet(cell_range: str, multiple_rows=False):
...
+        utils.log_error()
+        db.sheet_writes.set(utils.log_timestamp(), {'status': 'ERROR', 'log': sheet_log})
+
+
+def check_write_permission(spreadsheet: str, cell_range: str):
+    result = sheet.values().get(spreadsheetId=spreadsheet, range=cell_range).execute().get('values')
+    sheet.values().update(spreadsheetId=spreadsheet, range=cell_range, valueInputOption='USER_ENTERED',
+                          body={'values': result}).execute()
+
+re_column_row = re.compile(r"([A-Z]+)(\d+)$")
```
Comment (Kataiser) [inline spreadsheet.py]: This should be at the bottom, by `re_ping`

File: spreadsheet.py [hunk 6]
```diff
@@ -394,9 +442,11 @@ class SheetReadError(Exception):
...
 log: Optional[logging.Logger] = None
-SHEET_ID = '1yXTxFyIbqxjuzRt7Y8WCojpX2prULcfgiCZm1hWMbjE'
+SJ_SHEET_ID = '1yXTxFyIbqxjuzRt7Y8WCojpX2prULcfgiCZm1hWMbjE'
 creds = service_account.Credentials.from_service_account_file('service.json', scopes=['https://www.googleapis.com/auth/spreadsheets'])
 sheet = build('sheets', 'v4', credentials=creds).spreadsheets()
 difficulties = ("Beginner", "Intermediate", "Advanced", "Expert", "Grandmaster")
 re_ping = re.compile(r'<@\d+>')
 sj_data, sj_data_filenames = utils.load_sj_data()
+
+service_account_email: str = creds.service_account_email
```
Comment (Kataiser) [inline spreadsheet.py:453]: What is this for?

**Review by Kataiser — CHANGES_REQUESTED** [review]

File: spreadsheet.py (hunk 6 above)
Comment (jakobhellermann) [reply spreadsheet.py:453]: It's used in the setup command in the error for when the bot cannot access the google sheet.

```
Bot can't write to to https://docs.google.com/spreadsheets/d/1xY9W_fvKyYYz7E-t_t5UXSpqxECUil2mLY7PdB-ifLc at C2.
Make sure to invite bot-677@celestetasimprovementstracker.iam.gserviceaccount.com to the sheet.
```

**Comment by jakobhellermann** [comment]
I moved the setup from the `register_project` command into a separate `link_lobby_sheet` command that can be used like this:
- `link_lobby_sheet Testing https://docs.google.com/spreadsheets/d/1xY9W_fvKyYYz7E-t_t5UXSpqxECUil2mLY7PdB-ifLc C3`
- `link_lobby_sheet Testing https://docs.google.com/spreadsheets/d/1xY9W_fvKyYYz7E-t_t5UXSpqxECUil2mLY7PdB-ifLc/edit#gid=0 "Beginner Lobby!C3"`
- `link_lobby_sheet Testing 1xY9W_fvKyYYz7E-t_t5UXSpqxECUil2mLY7PdB-ifLc A0`

Do you have any preferences for how you would unlink the sheet? `unlink_lobby_sheet testing`? `link_lobby_sheet Testing none`?

File: spreadsheet.py (hunk 6 above)
Comment (Kataiser) [reply spreadsheet.py:453]: Oh ok, I couldn't find the reference to it for some reason

**Comment by Kataiser** [comment]
I don't need sheet unlinking support, people can just ask me if they need that

File: commands.py [hunk 7]
```diff
@@ -265,9 +267,61 @@ async def command_register_project(message: discord.Message, message_split: List
...
+    spreadsheet_id = sheet_match[1]
+
+    sheet_cell=message_split[3].strip('"')
+
+    lobby_sheet_cell = f"{spreadsheet_id}/{sheet_cell}"
+    spreadsheet_url=f"https://docs.google.com/spreadsheets/d/{spreadsheet_id}"
+
+    projects = db.projects.get_by_name_or_id(project_search_name)
+
+    if len(projects) == 0:
```
Comment (Kataiser) [inline commands.py]: Can just do `if not projects`

File: commands.py [hunk 8]
```diff
@@ -265,9 +267,61 @@ async def command_register_project(message: discord.Message, message_split: List
         await message.channel.send("Successfully verified and edited project.")
     else:
         add_mods_text = " Since you are doing sync checking, be sure to add mods (if need be) with the command `add_mods`." if do_sync_check.lower() == 'y' else ""
+        lobby_sheet_text = " If you want to automatically update lobby connection times on a google sheet, run `link_lobby_sheet`." if registered_project['is_lobby'] else ""
         await message.channel.send("Successfully verified and added project! If you want to change your project's settings, "
-                                   f"run the command again and it will overwrite what was there before.{add_mods_text}")
+                                   f"run the command again and it will overwrite what was there before.{add_mods_text}{lobby_sheet_text}")
 
+@command(re.compile(r'(?i)link_lobby_sheet .+ .+ (.+!)?[A-Z]+\d+'))
```
Comment (Kataiser) [inline commands.py]: This should be a reported command, add `report_usage=True`

File: commands.py
Comment (Kataiser) [inline commands.py:1]: There are a bunch of formatting conflicts vs what I do, but I'll just fix them myself

**Review by Kataiser — CHANGES_REQUESTED** [review]

**Pull request merged**